### PR TITLE
(4/4) Implement MRF Backend

### LINKED ISSUE #63: Robot Communication over Radio
**Is your feature request or enhancement related to a problem? What specifically could be better? Please describe.**
We kind of need an MRF backend to communicate with the physical robots.

**Describe the solution you'd like**
Port the MRF backend from the old code.

**Describe alternatives you've considered**
We could also rewrite it from scratch, but that would likely take a long time. 

**Additional context**
Breaking this down a bit:
- There is a libusb wrapper in `util/` that we should consider porting too. Additional classes in `util/` such as `NonCopyable` may also be useful to port over. Need to figure out where to put these.
- `Annunciator` is used for conveying messages from the field (ex. dead robots, hardware issues, basically whatever makes the dongle beep); this may have to be ported too and integrated into whatever logging system we come up with
- The old backend directly communicates with a `MRFRobot` instance; since we no longer have this and instead are using primitives only, this will have to be changed.
- `MRFRobot` has the packet encoding function, and some flag (`drive_dirty`) for `dirty_drive`, and keeps track of the charger state.
- It seems like `dirty_drive` is used to keep track of how many robots actually send primitives to the dongle, so it can adjust the packet size accordingly??? (`MRFDongle::submit_drive_transfer`)

### PR DESCRIPTION
**Pull request by henryymliu**
### Description
Adds a backend that sends primitives over radio.
The diff is large because it contains changes from #333 and #316, this should reduce after those PRs are merged.

### Testing Done
Field testing! Relies on legacy functions for encoding, which are not easily testable.

### Resolved Issues

Resolves #63 
### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by garethellis0** [comment]
I'm going to hold off on reviewing until #333 and #316 are merged if that's cool? It's just very tedious to figure out what's part of this PR vs those PR's.

File: src/thunderbots/software/launch/ai_mrf.launch [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    <node name="radio_communication" pkg="thunderbots" type="radio_communication" output="screen">
+    </node>
+
+    <!-- Launch the dynamic parameters node -->
+    <node name="parameters" pkg="thunderbots" type="parameters" output="screen"/>
+
+    <!-- Launch the rqt_reconfigure gui -->
+    <node name="rqt_reconfigure" pkg="rqt_reconfigure" type="rqt_reconfigure" output="screen"/>
+
+</launch>
```
Comment (garethellis0) [inline src/thunderbots/software/launch/ai_mrf.launch:10]: I've got a quick way to unify the commonalities between this and `ai_grsim.launch`, but I think it's easier if I just open a quick PR for that.

File: src/thunderbots/software/radio_communication/main.cpp [hunk 2]
```diff
@@ -27,8 +32,32 @@ void primitiveUpdateCallback(const thunderbots_msgs::PrimitiveArray::ConstPtr& m
...
 
+void worldUpdateCallback(const thunderbots_msgs::World::ConstPtr& msg)
+{
+    thunderbots_msgs::World world_msg = *msg;
+
+    // Extract team and ball data
+    Team friendly_team = Util::ROSMessages::createTeamFromROSMessage(world_msg.friendly_team);
+    Ball ball = Util::ROSMessages::createBallFromROSMessage(world_msg.ball);
+
+    std::vector<std::tuple<uint8_t, Point, Angle>> detbots;
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/main.cpp]: `detbots`? Not sure what this means, but I think a more verbose name would help here.

File: src/thunderbots/software/radio_communication/mrf_backend.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+     * Sends the given primitives to the backend to control the robots
+     *
+     * @param primitives the list of primitives to send
+     */
+    void sendPrimitives(const std::vector<std::unique_ptr<Primitive>>& primitives);
+
+    /**
+     * Updates the detected robots from vision.
+     *
+     * @param ft a vector of tuples of {robot id, robot location, robot orientation}
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/mrf_backend.h]: could we use a slightly more verbose name? not really sure what `ft` stands for here.

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking good, just some small naming comments.

**Comment by garethellis0** [comment]
Opened https://github.com/henrymliu/UBC-Thunderbots-Software/pull/2 with some launch file merging things

File: src/thunderbots/software/radio_communication/main.cpp (hunk 2 above)
Comment (henryymliu) [reply src/thunderbots/software/radio_communication/main.cpp]: renamed

File: src/thunderbots/software/radio_communication/mrf_backend.h (hunk 3 above)
Comment (henryymliu) [reply src/thunderbots/software/radio_communication/mrf_backend.h]: renamed

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Pull request merged**